### PR TITLE
People/davidgamba/rmd 22827 get flaw by cweid

### PR DESCRIPTION
**Pull request by DavidGamba**
@tonytvo please review.

### CODE REVIEW

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy [hunk 1]
```diff
@@ -142,7 +142,7 @@ class VeracodeDetailedReport {
     static List<Node> filterOpenFlaws(List<Node> flaws) {
         flaws.findAll { flaw ->
             String status = flaw.attribute('remediation_status')
-            (status == "Open" || status == "New" || status != "Fixed")
+            (status == "Open" || status == "New" || (status != "Fixed" && status != "Mitigated"))
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy:145]: what does this Mitigated mean? openFlawsAndMitigated?

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy [hunk 2]
```diff
@@ -276,4 +276,46 @@ class VeracodeDetailedReport {
     static List<List<String>> getOpenFlawRowsFromDetailedReport(Node xml) {
         return getFlawsAsRows(VeracodeDetailedReport.getOpenFlawsFromDetailedReportXML(xml))
     }
+
+    static void printFlawListByCWEID(Node xml, String cweid) {
+        List<Node> flaws = VeracodeDetailedReport.getOpenFlawsFromDetailedReportXML(xml)
+        println "issueid, remediation_status, mitigation_status, module, sourcefilepath, sourcefile, line, type"
+        flaws.each { flaw ->
+            String flawCWEID = flaw.attribute('cweid')
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy]: sounds like you can use filter to filter the flaws with cweid here

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy (hunk 1 above)
Comment (DavidGamba) [reply src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy:145]: Mitigated means that the flaw found by Veracode has counter measures in place that makes it not exploitable. So the flaw doesn't need to show up in the Open flaw list.

The name still applies to Open flaws, however, I made it so it shows the flaws that don't have a known status so I can catch new statuses that are not accounted for, that is how the Mitigated status showed up in my report when before I only knew about Fixed, Open and New.

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy [hunk 3]
```diff
@@ -276,4 +276,46 @@ class VeracodeDetailedReport {
...
+        // }
+        Map<String, Map<String, Object>> flawsByCWE = new HashMap<>()
+        flaws.each { flaw ->
+            String cweid = flaw.attribute('cweid')
+            List<String> attributes = XMLIO.getNodeAttributes(flaw, 'severity', 'categoryname')
+            Map<String, Object> m = flawsByCWE.get(cweid)
+            if (!m) {
+                m = new HashMap<>()
+                m.put('attributes', attributes)
+                m.put('count', 1)
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy]: sounds like you could extract the responsibilities of flaws counting out of this loop by using some kind of filter/map

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy [hunk 4]
```diff
@@ -276,4 +276,46 @@ class VeracodeDetailedReport {
...
+                flawsByCWE.put(cweid, m)
+                return
+            }
+            Integer count = m.get('count') as Integer
+            m.put('count', count + 1)
+            flawsByCWE.put(cweid, m)
+        }
+        println "CWEID, Severity, Count, Name"
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy]: sounds like you can extract the below into a function

**Review by tonytvo — APPROVED** [review]
it would be nice if we could keep the responsibilities inside the loop as low as possible
and again be careful with static function

**Pull request merged**